### ISSUE TITLE
Lag when typing in v2

### ISSUE THREAD

**Issue by jwhitmarsh** [dbcli/pgcli#979]
## Description
It's hard to demonstrate, but I have noticable lag when typing out a query in v2.0.1 compared to v1.11.0. It's usually at the end of a query, so i'll get as far as `update table set colum = value where...` and then it will start stuttering. Please let me know what information would be helpful to diagnose the problem.

## Your environment

- [ ] Please provide your OS and version information.
- [ ] Please provide your CLI version.
- [ ] What is the output of ``pip freeze`` command.

MacOS `10.13.6`
pgcli `v2.0.1`

```
➜ pip freeze
awscli==1.16.13
boto3==1.7.70
botocore==1.12.3
certifi==2018.8.24
chardet==3.0.4
cli-helpers==1.1.0
Click==7.0
colorama==0.3.9
configobj==5.0.6
docutils==0.14
httpie==0.9.9
httpie-jwt-auth==0.3.0
humanize==0.5.1
idna==2.7
jmespath==0.9.3
pgcli==1.11.0
pgspecial==1.11.4
prompt-toolkit==1.0.15
psycopg2==2.7.6.1
pyasn1==0.4.4
Pygments==2.2.0
python-dateutil==2.7.3
python-magic==0.4.15
pytz==2018.5
PyYAML==3.13
requests==2.19.1
rsa==3.4.2
s3cmd==2.0.2
s3transfer==0.1.13
setproctitle==1.1.10
six==1.11.0
sqlparse==0.2.4
tabulate==0.8.2
terminaltables==3.1.0
urllib3==1.23
wcwidth==0.1.7
You are using pip version 18.0, however version 18.1 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

**Comment by j-bennet** [comment]
I think it's most likely a prompt-toolkit issue. Migrating to prompt-toolkit 2.0 was the only change in `pgcli` 2.0. @jonathanslenders can you advise?

**Comment by jonathanslenders** [comment]
One change in prompt_toolkit 2.0 is that completions are now synchronous by default. If they are expensive, they have to be wrapped into ``prompt_toolkit.completion.ThreadedCompleter``, otherwise they'll block the event loop and make the UI unresponsive. Could this be the cause?

**Comment by amjith** [comment]
Is there a reason for synchronous completion being the default?

**Comment by j-bennet** [comment]
> If they are expensive, they have to be wrapped

By "expensive" you mean they take long to retrieve? Or there's a lot of them? From the timings that I logged, it seems that retrieving those values is fast (less than a millisecond), but is it possible that rendering a lot is problematic? In pgcli, we seem to be displaying ~3k items in autocomplete (mostly function names).

**Comment by hunsakerbn** [comment]
I've also seen this lag. It is very annoying.

**Comment by loisch** [comment]
For me pgcli is unusable right now.  I'm experience typing pauses in the range of seconds - even when there are only a few completions.

**Comment by j-bennet** [comment]
@loisch your case sounds like the worst one. Can you please provide a few more details: your OS, what plugins are installed in Postgres (`\dx`), and if you're using a VM?

**Comment by jonathanslenders** [comment]
Did you try to wrap this completer into ``ThreadedCompleter`` anyone?

See this line:
https://github.com/dbcli/pgcli/blob/master/pgcli/main.py#L729

Change it with:
```python
from prompt_toolkit.completion import ThreadedCompleter
....
completer=ThreadedCompleter(DynamicCompleter(lambda: self.completer)),
```

Are you doing any network I/O for generating the completions?

**Comment by j-bennet** [comment]
@jonathanslenders There's a few factors in play here:

* we have to query the database to retrieve metadata (tables, columns, functions). This could cause a slowdown with a remote database. If this is the problem, `ThreadedCompleter` could help.
* compared to `mycli` (for example), `pgcli` may end up pulling a ton of function names into the list of completions. Lots of those come from installed extensions. It is possible that we should exclude those from completions, as their usefulness is doubtful.

I'm trying to figure out what it can be, and then we can we start trying fixes. I also need a reliable way to reproduce the problem. I've seen the slowdown, but it was not very pronounced, and only appeared with a huge list of completions.

**Comment by j-bennet** [comment]
For those affected by the lag problem, can you install `pgcli` from branch:

`pip install git+https://github.com/dbcli/pgcli.git@j-bennet/threaded-completer`

and see if this improves things for you? Please give a 👍 or 👎 here.

**Comment by chocobn69** [comment]
It's better for me, but not as smooth as it was before prompt-toolkit 2 upgrade

**Comment by j-bennet** [comment]
@loisch Very interesting! If you run into this problem again, please do not kill your history file, it could be useful to us to debug what's going on (unless you cannot share it for privacy reasons). I'm glad the problem is remedied for now. I'm still working on another, hopefully more complete fix.

**Comment by j-bennet** [comment]
After some more digging, the lag is reproducible (and very pronounced) only when using multiline mode. Otherwise, things are quite speedy. Here is the multiline query I was using:

```
SELECT n.nspname as "Schema",
  p.proname as "Name",
  pg_catalog.pg_get_function_result(p.oid) as "Result data type",
  pg_catalog.pg_get_function_arguments(p.oid) as "Argument data types",
 CASE
  WHEN p.proisagg THEN 'agg'
  WHEN p.proiswindow THEN 'window'
  WHEN p.prorettype = 'pg_catalog.trigger'::pg_catalog.regtype THEN 'trigger'
  ELSE 'normal'
 END as "Type"
FROM pg_catalog.pg_proc p
     LEFT JOIN pg_catalog.pg_namespace n ON n.oid = p.pronamespace
WHERE pg_catalog.pg_function_is_visible(p.oid)
      AND n.nspname <> 'pg_catalog'
      AND n.nspname <> 'information_schema'
ORDER BY 1, 2, 4;
```

In multiline mode, I see the lag:

* even with the `ThreadedCompleter`
* even when querying a local database.

So here is another debugging question to everyone affected: do you see the lag only/mostly in multiline mode (click 👍 ), or in single line mode also (click 👎 )?

**Comment by chocobn69** [comment]
I tried the `remove-some-functions-completions` and I can say it's almost as fast as it was before ! Thanks !

**Comment by j-bennet** [comment]
Both fixes has been merged and released as part of `pgcli` v2.0.2. Please update with:

`pip install -U pgcli`
